Mathlib/RingTheory/FilteredAlgebra/Basic.lean

```diff
@@ -0,0 +1,124 @@
+
+In this file, we defined the fitration on abelian group,
+and extend it to get the filtration on ring.
+
+# Main definitions
+
+* `IsFiltration` : For `σ` satisfying `SetLike σ A`, an increasing series of `F` in `σ`
+is filtration if there is another series `F_lt` equal to the supremum of `F` with smaller index
+
+* `IsRingFiltration` : For `σ` satisfying `SetLike σ R` where `R` is a semiring,
+an increasing series `F` in `σ` is ring filtration if `IsFiltration F F_lt` and
+the pointwise multiplication of `F i` and `F j` is in `F (i + j)`
+
+* `IsModuleFiltration` : For `F` satisfying `IsRingFiltration F F_lt` in a semiring `R` and `σM`
+satisfying `SetLike σ M` where `M` is a module over `R`, an increasing series `FM` in `σM` is
+module filtration if `IsFiltration F F_lt` and the pointwise scalar multiplication of
+`F i` and `FM j` is in `F (i +ᵥ j)`
+
+-/
+
+section GeneralFiltration
+
+variable {ι : Type*} [Preorder ι]
+
+variable {A : Type*} {σ : Type*} [SetLike σ A]
+
+/--For `σ` satisfying `SetLike σ A`, an increasing series of `F` in `σ` is filtration if
+there is another series `F_lt` equal to the supremum of `F` with smaller index.
+
+In fact `F_lt j = ⨆ i < j, F i`, the design of `F_lt` can handle different conditions in the
+same structure, it avoid adding `CompleteLattice` to `σ`, also providing convenience when the index
+is `ℤ` -/
+class IsFiltration (F : ι → σ) (F_lt : outParam <| ι → σ) : Prop where
+  mono : Monotone F
+  is_le {i j} : i < j → F i ≤ F_lt j
+  is_sup (B : σ) (j : ι) : (∀ i < j, F i ≤ B) → F_lt j ≤ B
+
+/--A special case of `IsFiltration` when index is integer-/
+lemma IsFiltration_int (F : ℤ → σ) (mono : Monotone F) :
+    IsFiltration F (fun n ↦ F (n - 1)) where
+  mono := mono
+  is_le lt := mono (Int.le_sub_one_of_lt lt)
+  is_sup _ j hi := hi (j - 1) (sub_one_lt j)
+
+end GeneralFiltration
+
+section FilteredRing
+
+variable {ι : Type*} [OrderedAddCommMonoid ι]
+
+variable {R : Type*} [Semiring R] {σ : Type*} [SetLike σ R]
+
+/--For `σ` satisfying `SetLike σ R` where `R` is a semiring, an increasing series `F` in `σ` is ring
+filtration if `IsFiltration F F_lt` and the pointwise multiplication of `F i` and `F j` is in
+`F (i + j)`-/
+class IsRingFiltration (F : ι → σ) (F_lt : outParam <| ι → σ)
+    extends IsFiltration F F_lt : Prop where
+  one_mem : 1 ∈ F 0
+  mul_mem : ∀ {i j x y}, x ∈ F i → y ∈ F j → x * y ∈ F (i + j)
+
+instance [AddSubmonoidClass σ R] (F : ι → σ) (F_lt : outParam <| ι → σ) [IsRingFiltration F F_lt] :
+    Semiring (F 0) where
+  mul x y :=  ⟨x.1 * y.1, by simpa using IsRingFiltration.mul_mem x.2 y.2⟩
+  left_distrib a b c := SetCoe.ext (mul_add a.1 b.1 c.1)
+  right_distrib a b c := SetCoe.ext (add_mul a.1 b.1 c.1)
+  zero_mul a := SetCoe.ext (zero_mul a.1)
+  mul_zero a := SetCoe.ext (mul_zero a.1)
+  mul_assoc a b c := SetCoe.ext (mul_assoc a.1 b.1 c.1)
+  one := ⟨1, IsRingFiltration.one_mem⟩
+  one_mul a := SetCoe.ext (one_mul a.1)
+  mul_one a := SetCoe.ext (mul_one a.1)
```

Comment:
With the change above, you can delete this instance entirely, as it's just [`SetLike.GradeZero.instSemiring`](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Algebra/DirectSum/Internal.html#SetLike.GradeZero.instSemiring).